contracts/Pool.sol

```diff
@@ -0,0 +1,381 @@
+   * The borrowers registry decides if an account can borrow funds.
+   * Only the owner of the Contract can execute this function.
+   * @dev _borrowersRegistry the address of borrowers registry
+  **/
+  function setBorrowersRegistry(IBorrowersRegistry borrowersRegistry_) external onlyOwner {
+    require(address(borrowersRegistry_) != address(0), "The borrowers registry cannot set to a null address");
+
+    _borrowersRegistry = borrowersRegistry_;
+  }
+
+
+  /* ========== MUTATIVE FUNCTIONS ========== */
+  function transfer(address recipient, uint256 amount) external override returns (bool) {
+    require(recipient != address(0), "ERC20: cannot transfer to the zero address");
+    require(recipient != address(this), "ERC20: cannot transfer to the pool address");
+
+    _accumulateDepositInterest(msg.sender);
+
+    require(_deposited[msg.sender] >= amount, "ERC20: transfer amount exceeds balance");
+
+    // (this is verified in "require" above)
+    unchecked {
+      _deposited[msg.sender] -= amount;
+    }
+
+    _accumulateDepositInterest(recipient);
+    _deposited[recipient] += amount;
+
+    emit Transfer(msg.sender, recipient, amount);
+
+    return true;
+  }
+
+  function allowance(address owner, address spender) external view override returns (uint256) {
+    return _allowed[owner][spender];
+  }
+
+  function approve(address spender, uint256 amount) external override returns (bool) {
+    _accumulateDepositInterest(msg.sender);
+
+    require(_deposited[msg.sender] >= amount, "ERC20: approve amount exceeds balance");
+    _allowed[msg.sender][spender] = amount;
+
+    emit Approval(msg.sender, spender, amount);
+
+    return true;
+  }
+
+  function transferFrom(address sender, address recipient, uint256 amount) external override returns (bool) {
+    require(amount <= _allowed[sender][msg.sender], "Not enough tokens allowed to transfer required amount.");
+    require(recipient != address(0), "ERC20: cannot transfer to the zero address");
+    require(recipient != address(this), "ERC20: cannot transfer to the pool address");
+
+    _accumulateDepositInterest(msg.sender);
+
+    require(amount <= _deposited[sender], "Not enough tokens to transfer required amount.");
+
+    _deposited[sender] -= amount;
+    _allowed[sender][msg.sender] -= amount;
+
+    _accumulateDepositInterest(recipient);
+    _deposited[recipient] += amount;
+
+    emit Transfer(sender, recipient, amount);
+
+    return true;
+  }
+
+
+  /**
+   * Deposits the message value
+   * It updates user deposited balance, total deposited and rates
+  **/
+  function deposit() payable virtual external {
+    _accumulateDepositInterest(msg.sender);
+
+    _mint(msg.sender, msg.value);
+    _updateRates();
+
+    emit Deposit(msg.sender, msg.value, block.timestamp);
+  }
+
+  /**
+   * Withdraws selected amount from the user deposits
+   * @dev _amount the amount to be withdrawn
+  **/
+  function withdraw(uint256 _amount) external nonReentrant {
```

Comment:
currently it misses a basic check- whether pool account balance is enough for withdraw. I think we should do it before any other action, @jakub-wojciechowski what do you think?

Reply:
It will fail in the transfer but I agree it'll be a good practice to return a meaningful error message. It'd be great if you could add it @piotruda